### PR TITLE
Enhance reward token balance discovery

### PR DESCRIPTION
**Pull request by tintinthong**
Reward Token Balance
Screenshot 2021-09-29 at 9 41 30 PM

Merkle Proofs
Screenshot 2021-09-29 at 9 41 02 PM

Demo here

https://www.loom.com/share/ea24cc2cd76f4341b8d2cf1d5398e80d

### CODE REVIEW

**Comment by tintinthong** [comment]
WIP. shall refactor this.

**Comment by habdelra** [comment]
If you use prettier to format on save you shouldn't have to worry about the linting failure that you ran into: consider using this https://github.com/prettier/prettier-emacs (I know you're an emacs fan)

File: packages/cardpay-sdk/contracts/addresses.ts [hunk 1]
```diff
@@ -89,6 +95,9 @@ const XDAI = {
   uniswapV2Router: '0x1C232F01118CB8B424793ae03F870aa7D0ac7f77', // This is the UniswapV2Router02
   uniswapV2Factory: '0xA818b4F111Ccac7AA31D0BCc0806d64F2E0737D7',
   rewardPool: '0x42628325845B3e01EfdD6ce4b3665453dC6c9A13',
+  // rewardManager: '',
+  // registerRewardProgramHandler: '',
+  // registerRewardeeHandler: '',
```
Comment (habdelra) [inline packages/cardpay-sdk/contracts/addresses.ts:100]: I mentioned it in another PR, but don't comment these out--just leave them in there with empty values. If we forget to set these addresses, the resulting error messages will be very useful for reminding us to do this.

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 2]
```diff
@@ -360,6 +450,17 @@ The reward program ${rewardProgramId} has balance equals ${fromWei(
...
+    return tokenAddresses.reduce(async (obj, tokenAddress: string) => {
+      const tokenContract = new this.layer2Web3.eth.Contract(ERC20ABI as AbiItem[], tokenAddress);
+      const tokenSymbol = await tokenContract.methods.symbol().call();
+      return {
+        ...obj,
+        [tokenAddress]: tokenSymbol,
+      };
+    }, {});
+  }
+  // tokenMapping: any =
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:463]: can you please cleanup the commented code

**Comment by tintinthong** [comment]
Previously this branch had some subgraph stuff which are not relevant to it. I have removed that.

File: packages/cardpay-cli/reward-pool.ts [hunk 3]
```diff
@@ -13,24 +13,16 @@ export async function rewardTokenBalances(
...
+  console.log(`Reward balances for ${address}`);
+  displayRewardTokenBalance(tokenBalances);
 }
 
 function displayRewardTokenBalance(tokenBalances: RewardTokenBalance[]): void {
   const groupedByRewardProgram = groupBy(tokenBalances, (a) => a.rewardProgramId);
   Object.keys(groupedByRewardProgram).map((rewardProgramId: string) => {
+    console.log('---------------------------------------------------------------------');
     console.log(`    RewardProgram: ${rewardProgramId}`);
     console.log('---------------------------------------------------------------------');
```
Comment (habdelra) [inline packages/cardpay-cli/reward-pool.ts:27]: if you use a ` then you can wrap new lines without having to repeat the console.log statements like this:
```js
console.log(`
---------------------------------------------------------------------
RewardProgram: ${rewardProgramId}
---------------------------------------------------------------------`);

File: packages/cardpay-cli/reward-pool.ts [hunk 4]
```diff
@@ -13,24 +13,16 @@ export async function rewardTokenBalances(
...
-  } else {
-    const tokenBalances = await rewardPool.rewardTokenBalances(address);
-    console.log('\n');
-    console.log(`Reward balances for ${address}`);
-    console.log('---------------------------------------------------------------------');
-    displayRewardTokenBalance(tokenBalances);
-  }
+  const tokenBalances = await rewardPool.rewardTokenBalances(address, rewardProgramId);
+  console.log('\n');
+  console.log(`Reward balances for ${address}`);
```
Comment (habdelra) [inline packages/cardpay-cli/reward-pool.ts:18]: anytime you have `console.log('\n');` that probably means you should just use `s like this:
```js
console.log(`
Reward balances for ${address}`)'
```

File: packages/cardpay-cli/reward-pool.ts [hunk 5]
```diff
@@ -41,13 +33,19 @@ function displayRewardTokenBalance(tokenBalances: RewardTokenBalance[]): void {
...
+    console.log('---------------------------------------------------------------------');
+    console.log(`    RewardProgram: ${rewardProgramId}`);
+    console.log('---------------------------------------------------------------------');
+    let p = groupedByRewardProgram[rewardProgramId];
+    p.map((o) => {
+      console.log(`  rewardProgramId: ${o.rewardProgramId}`);
+      console.log(`  token: ${o.tokenAddress} (${o.tokenSymbol})`);
+      console.log(`  balance: ${fromWei(o.balance)}`);
+      console.log(`  paymentCycle: ${o.paymentCycle}`);
+      console.log(`  proof: ${o.proof}`);
```
Comment (habdelra) [inline packages/cardpay-cli/reward-pool.ts:47]: you can rewrite this whole thing with a single `console.log`:
```js
console.log(`
  rewardProgramId: ${o.rewardProgramId}`);
  token: ${o.tokenAddress} (${o.tokenSymbol})
  balance: ${fromWei(o.balance)}
  paymentCycle: ${o.paymentCycle}
  proof: ${o.proof}`);
```

File: packages/cardpay-cli/reward-pool.ts [hunk 6]
```diff
@@ -99,8 +97,7 @@ export async function getClaimableRewardProofs(
   let rewardPool = await getSDK('RewardPool', web3);
   const claimableRewardProofs = await rewardPool.getProofsWithNonZeroBalance(address, rewardProgramId, tokenAddress);
   console.log('\n');
-  console.log(`Proofs balances for ${address}`);
-  console.log('--------------------------------------------------------------');
+  console.log(`Reward Proofs for ${address}`);
```
Comment (habdelra) [inline packages/cardpay-cli/reward-pool.ts:100]: better written as:
```js
console.log(`
Reward Proofs for ${address}`);
```

File: packages/cardpay-cli/reward-pool.ts [hunk 7]
```diff
@@ -99,8 +97,7 @@ export async function getClaimableRewardProofs(
   let rewardPool = await getSDK('RewardPool', web3);
   const claimableRewardProofs = await rewardPool.getProofsWithNonZeroBalance(address, rewardProgramId, tokenAddress);
   console.log('\n');
```
Comment (habdelra) [inline packages/cardpay-cli/reward-pool.ts:99]: remove this and just use `s int eh console.log

**Review by habdelra — APPROVED** [review]
👍

**Pull request merged**